session/session.go

```diff
@@ -96,13 +96,22 @@ func newZKSession(servers string, recvTimeout time.Duration, logger stdLogger, c
 	}
 
 	var event zookeeper.Event
-	select {
-	case event = <-events:
-		if event.State != zookeeper.STATE_CONNECTED {
+	for {
+		select {
+		case event = <-events:
+			switch event.State {
+				case zookeeper.STATE_AUTH_FAILED:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_EXPIRED_SESSION:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_CONNECTED:
+					break
```

Comment:
Can you explain these states? Why are we guaranteed they will be the first events sent on the connection?